NewsApp/Controllers/WeatherViewController/WeatherViewController.swift

```diff
@@ -0,0 +1,438 @@
+     
+     */
+
+    
+    private func setupSearchController() {
+        searchController = UISearchController(searchResultsController: nil)
+        searchController.searchResultsUpdater = self
+        searchController.obscuresBackgroundDuringPresentation = false
+        searchController.searchBar.placeholder = "Search for a city"
+        navigationItem.searchController = searchController
+        definesPresentationContext = true
+    }
+    
+    private func setupLocationManager() {
+        locationManager.delegate = self
+        locationManager.desiredAccuracy = kCLLocationAccuracyBest
+        locationManager.requestWhenInUseAuthorization()
+        locationManager.startUpdatingLocation() // Start fetching the location
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        if let location = locations.first {
+            locationManager.stopUpdatingLocation()
+            
+            let lat = location.coordinate.latitude
+            let lon = location.coordinate.longitude
+            
+            print("Location coordinates fetched: lat = \(lat), lon = \(lon)")
+            let geocoder = CLGeocoder()
+            geocoder.reverseGeocodeLocation(location) { [weak self] (placemarks, error) in
+                if let error = error {
+                    print("Reverse geocoding error: \(error.localizedDescription)")
+                    self?.showError("Unable to fetch city name")
+                    return
+                }
+                if let placemark = placemarks?.first, let city = placemark.locality {
+                    // Update the cityLabel with the detected city name
+                    print("Detected city name: \(city)") // Debug statement to check city
+                    DispatchQueue.main.async {
+                        self?.cityLabel.text = city
+                    }
+                    // Fetch weather data using the detected coordinates
+                    self?.fetchWeatherData(for: lat, lon: lon)
+                } else {
+                    self?.showError("City name not found")
+                }
+            }
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        print("Failed to get user location: \(error.localizedDescription)")
+        showError("Unable to get location. Please enable location services.")
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        if status == .denied || status == .restricted {
+            // If location permission is denied, show an error
+            showError("Location services are disabled. Please enable them in settings.")
+        } else if status == .authorizedWhenInUse || status == .authorizedAlways {
+            // Location access was granted, start updating location
+            locationManager.startUpdatingLocation()
+        }
+    }
```

Comment:
To reduce the code in your viewController, you can move al lthese location related functions to an extension.